source/cli/ArgumentParser.ooc

```diff
@@ -0,0 +1,109 @@
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this software. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use ooc-collections
+use ooc-base
+import Event
+
+Argument: class {
+	_longIdentifier: Text
+	_shortIdentifier: Text
+	_parameters: Int
+	_action1: Event
+	_action2: Event1<Text>
+	_action3: Event1<VectorList<Text>>
+	init: func (=_longIdentifier, =_shortIdentifier, =_parameters, =_action1)
+	init: func ~parameter (=_longIdentifier, =_shortIdentifier, =_parameters, =_action2)
+	init: func ~parameters (=_longIdentifier, =_shortIdentifier, =_parameters, =_action3)
+	init: func ~default (=_action2)
+	free: override func {
+		if (this _action1 != null)
+			this _action1 free()
+		if (this _action2 != null)
+			this _action2 free()
+		if (this _action3 != null)
+			this _action3 free()
+		super()
+	}
+}
+
+ArgumentParser: class {
+	_arguments: VectorList<Argument>
+	_defaultArgument: Argument
+	init: func {
+		this _arguments = VectorList<Argument> new()
+	}
+	free: override func {
+		this _arguments free()
+		super()
+	}
+	add: func (longIdentifier: Text, shortIdentifier: Text, parameters: Int, action: Event) {
+		this _arguments add(Argument new(longIdentifier, shortIdentifier, parameters, action))
+	}
+	add: func ~parameter (longIdentifier: Text, shortIdentifier: Text, parameters: Int, action: Event1<Text>) {
+		this _arguments add(Argument new(longIdentifier, shortIdentifier, parameters, action))
+	}
+	add: func ~parameters (longIdentifier: Text, shortIdentifier: Text, parameters: Int, action: Event1<VectorList<Text>>) {
+		this _arguments add(Argument new(longIdentifier, shortIdentifier, parameters, action))
+	}
+	addDefault: func (action: Event1<Text>) {
+		this _defaultArgument = Argument new(action)
+	}
+	parse: func (input: VectorList<Text>) {
+		identifier: Text
+		argument: Argument
+		parameters := VectorList<Text> new()
+		for (i in 0 .. input count) {
+			identifier = t""
+			argument = null
+			parameters clear()
+			if (input[i] beginsWith(t"--"))
+				identifier = input[i] slice(2, input[i] count-2)
+			else if (input[i] beginsWith(t"-"))
+				identifier = input[i] slice(1, input[i] count-1)
+			else {
+				_defaultArgument _action2 call(input[i])
+			}
+			argument = _findArgument(identifier)
+			if (argument) {
+				if (argument _parameters > 1) {
+					for (j in 0 .. argument _parameters)
+						parameters add(input[i + j + 1])
+					argument _action3 call(parameters)
+					i = i + argument _parameters
+				}
+				else if (argument _parameters == 1) {
+					argument _action2 call(input[i + 1])
+					i = i + 1
+				}
+				else
+					argument _action1 call()
+			}
+		}
+		parameters free()
+	}
+	_findArgument: func (identifier: Text) -> Argument{
```

Comment:
Magic should catch this, @thomasfanell 
Anyway, you're missing a space before `{`